doc/wordlint.1
.TH "WORDLINT" "1" "2014\-11\-22" "0.1.0.0
.SH Name
.PP
wordlint \- plaintext redundancy linter
.SH Synopsis
.PP
wordcheck [OPTIONS] [\-f input\-file]
.SH Description
.PP
Wordlint locates matching pairs of words repeated within a user\-defined
distance.
Text may be linted by distance between words (that is, by word count),
by line count, or by percentage of the total words in the file.
The user may also choose a minimum word length for matches.
.PP
Options exist for data output, which is machine\-readable by default.
Results may be sorted by alphabetically by word, by position (line
number), or by intervening distance between matches; and may be used
with a human\-readable mode.
Additionally, a Vim mode provides output for a plugin.
.SH Options
.TP
.B \-f \f[I]FILE\f[]
Specify an input file.
If none is given, wordlint reads from stdin.
.RS
.RE
.SS Linting Options
.TP
.B \-d \f[I]INT | FLOAT\f[]
Specify maximum intervening distance between returned word\-pairs.
\f[B]If the type of lint is either "word" or "line", an integer must be
used\f[], while a "percentage" check will accept a float value.
Ignored if \-a is used.
Default is 250.
.RS
.RE
.TP
.B \-t \f[I]word|line|percentage\f[]
Specify type of lint to perform, which affects which the calculation of
intervening distance between word pairs.
Options are:
.RS
.IP
.nf
\f[C]
\-\ word\ (default)
\-\ line
\-\ percentage
\f[]
.fi
.PP
A word\-type check will define a word\[aq]s "position" as it\[aq]s word
count, while a line check uses the line number on which a word is found.
A percentage check sets this value according to a word\[aq]s count
divided by the total count of words in the input.
.RE
.TP
.B \-w \f[I]NUMBER\f[]
Specify minimum length of words to be matched, i.e.
to reduce hits for "there".
Default is 5.
.RS
.RE
.SS Output Options
.TP
.B \-a
Return all matched pairs of words regardless of intervening distance.
Deactivates \-d parameter.
.RS
.RE
.TP
.B \-h
Return human\-readable output.
Compatible with all sorting except for vim.
.RS
.RE
.TP
.B \-s \f[I]word|position|distance|vim\f[]
Sort word pairs alphabetically, by line number, or by intervening
distance; or provides Vim plugin output respective to the following
options:
.RS
.IP
.nf
\f[C]
\-\ word
\-\ position\ (default)
\-\ distance
\-\ vim
\f[]
.fi
.RE
.SH Examples
.IP
.nf
\f[C]
wordlint\ \-t\ line\ \-d\ 20\ \-w\ 7\ \-f\ file.txt
\f[]
.fi
.PP
Finds matching strings consisting of seven characters or more and which
have an intervening distance of twenty or less.
.IP
.nf
\f[C]
cat\ file.txt\ |\ wordlint\ \-t\ percentage\ \-d\ 2.5\ \-a\ \-h\ \-s\ word
\f[]
.fi
.PP
Finds all matching, five\-character\-or\-more strings found within a
2.5% distance of one\-another, and returns the output in human\-readable
form, sorted alphabetically.
.SH See Also
.PP
A Vim front\-end to Wordlint, creatively named Wordlint.vim, is
available at https://github.com/gbgar/Wordlint.vim
.SH AUTHOR
GB Gardner.


changelog
2014-11-25 - version 0.1.0.1 
=============================

Added labels to machine-readable output table.

Minor documntation fixes


Wordlint.hs
import Data.List
import System.Console.CmdArgs
import Text.PrettyPrint.Boxes 
import Wordlint.Args
import Wordlint.Words
import Wordlint.Wordpairs

main :: IO ()
main = do
    -- Execute command line arguments, retrieve flags
    cargs <- cmdArgs cliargs
    let sortflag = sort_ cargs
    let wordlen = wordlength cargs
    -- If human-readable flag is present, print header
    checkIfHumanHeader cargs
    -- Acquire String data from file or stdin
    dat <- accessInputFileData . checkFileStdin $ file cargs
    -- blacklist data for running filter function
    blist' <- accessBlacklistFileData . checkFileStdin $ blacklist cargs
    let blist = setBlacklistData blist'
    -- 
    -- Choose checker and printer according to -t, -h flags
    -- 
    case type_ cargs  of
        [] -> do 
            let checkedwords' = runWordCheck dat wordlen dist blist cargs
            let checkedwords  = checkSortFlag sortflag checkedwords'
            if human cargs 
            then do putStrLn "No type chosen; running word-based check"
                    mapM_ putStrLn (processHumanWordData checkedwords)
                    putStrLn ""
                    summaryData (length checkedwords) (wordlength cargs) (length dat)
            else 
                if sortflag == "vim" 
                then putStrLn $ processDataVim checkedwords
                else processMachineWordData checkedwords
                             where dist = checkDistanceOrAll cargs
        "word" -> do
            let checkedwords' = runWordCheck dat wordlen dist blist cargs
            let checkedwords =  checkSortFlag sortflag checkedwords'

            if human cargs
            then do mapM_ putStrLn (processHumanWordData checkedwords) 
                    putStrLn ""
                    summaryData (length checkedwords) (wordlength cargs) (length dat) 
            else 
                if sortflag == "vim"
                then putStrLn $ processDataVim checkedwords
                else processMachineWordData checkedwords
                            where dist = checkDistanceOrAll cargs 
        "line" -> do
            let checkedwords' = runLineCheck dat wordlen dist blist cargs
            let checkedwords =  checkSortFlag sortflag checkedwords'
            if human cargs
            then do mapM_ putStrLn (processHumanLineData checkedwords)
                    putStrLn ""
                    summaryData (length checkedwords) (wordlength cargs) (length dat)
            else 
                if sortflag == "vim"
                then putStrLn $ processDataVim checkedwords
                else processMachineLineData checkedwords
                            where dist = checkDistanceOrAll cargs
        "percentage" -> do 
            let checkedwords' = runPercentageCheck dat wordlen dist blist cargs
            let checkedwords =  checkSortFlag sortflag checkedwords'
            if human cargs
            then do mapM_ putStrLn (processHumanPercentageData checkedwords)
                    putStrLn ""
                    summaryData (length checkedwords) (wordlength cargs) (length dat)
            else 
                if sortflag == "vim"
                then putStrLn $ processDataVim checkedwords
                else processMachinePercentageData checkedwords
                            where dist = checkDistanceOrAll cargs

        _ -> do 
            let checkedwords' = runWordCheck dat wordlen dist blist cargs
            let checkedwords =  checkSortFlag sortflag checkedwords'
            if human cargs 
            then do putStrLn "No type chosen; running word-based check"
                    mapM_ putStrLn (processHumanWordData checkedwords)
                    putStrLn ""
                    summaryData (length checkedwords) (wordlength cargs) (length dat)
            else 
                if sortflag == "vim" 
                then putStrLn $ processDataVim checkedwords
                else processMachineWordData checkedwords
                             where dist = checkDistanceOrAll cargs
-- run*Check functions are run with a maybe to account for --all flag
-- process*Data functions return a human-readable format. 
-- In the future, a "machine-readable" flag will be added to the latter
-- in order to output for use in text editor plugins.

runWordCheck :: String -> Int -> Maybe Int -> Maybe [String] -> Arguments -> Wordpairs Int 
runWordCheck inputdata wordlen distorall blist arg = case distorall of
    Nothing -> instring
    Just i -> filterWordpairsByDistance instring i
  where instring = sortWordsByString . filterMatchingWords $ dwords
        dwords = checkWordList fwords wordlen
        fwords = runFilterFlags cwords arg blist
        cwords = zipWords inputdata "word" 

processHumanWordData :: Wordpairs Int -> [String]
processHumanWordData [] = ["No (more) matches found"]
processHumanWordData (x:xs) = ("\'" ++ word ++ "\'"
                                    ++ " at positions "
                                    ++ position'
                                    ++ " at coordinates "
                                    ++ coordinates
                                    ++ " with an intervening distance of "
                                    ++ distance'
                                    ++ " words") : processHumanWordData xs
                         where word = getWordPairString x
                               position' = show (getWordpairPositions x)
                               coordinates = show (getWordpairCoords x)
                               distance' = show (pdiff x)

-- machine-based outputs drawn from
-- http://www.tedreed.info/programming/2012/06/02/how-to-use-textprettyprintboxes/
processMachineWordData :: Wordpairs Int -> IO ()
processMachineWordData x = printBox $ hsep 2 left (map (vcat left . map text)
                            (transpose $ ["Coord1", "Coord2", "Word", "Distance"]:processMachineWordData' x))



processMachineWordData' :: Wordpairs Int -> [[String]]
processMachineWordData' [] = []
processMachineWordData' (x:xs) = words (coordinates1 coordinates'
                                 ++ " "
                                 ++ coordinates2 coordinates'
                                 ++ " "
                                 ++ word 
                                 ++ " "
                                 ++ distance') : processMachineWordData' xs
                         where word = getWordPairString x
                               coordinates' = getWordpairCoords x
                               coordinates1 ((r1,s1),(_,_)) = show r1 ++ "," ++ show s1
                               coordinates2 ((_,_),(r2,s2)) = show r2 ++ "," ++ show s2
                               distance' = show (pdiff x)

runLineCheck :: String -> Int -> Maybe Int -> Maybe [String] -> Arguments-> Wordpairs Int
runLineCheck inputdata wordlen distorall blist arg = case distorall  of
    Nothing ->  instring 
    Just i -> filterWordpairsByDistance instring  i
  where instring = sortWordsByString . filterMatchingWords $ dwords
        dwords = checkWordList fwords wordlen
        fwords = runFilterFlags cwords arg blist
        cwords = zipWords inputdata "line" 

processHumanLineData :: Wordpairs Int -> [String]
processHumanLineData [] = ["No (more) matches found"]
processHumanLineData (x:xs) = ("\'" ++ word ++ "\'"
                         ++ " at coordinates "
                         ++ coordinates
                         ++ " with an intervening distance of "
                         ++ distance'
                         ++ " lines") : processHumanLineData xs
                         where word = getWordPairString x
                               coordinates = show (getWordpairCoords x)
                               distance' = show (pdiff x)
                               
processMachineLineData :: Wordpairs Int -> IO ()
processMachineLineData x = printBox $ hsep 2 left (map (vcat left . map text)
                            (transpose $ ["Coord1", "Coord2", "Word", "Distance"]:processMachineLineData' x))

processMachineLineData' :: Wordpairs Int -> [[String]]
processMachineLineData' [] = []
processMachineLineData' (x:xs) = words (coordinates1 coordinates' 
                                 ++ " "
                                 ++ coordinates2 coordinates'
                                 ++ " "
                                 ++ word 
                                 ++ " "
                                 ++ distance')
                                 : processMachineLineData' xs
                         where word = getWordPairString x
                               coordinates' = getWordpairCoords x
                               coordinates1 ((r1,s1),(_,_)) = show r1 ++ "," ++ show s1
                               coordinates2 ((_,_),(r2,s2)) = show r2 ++ "," ++ show s2
                               distance' = show (pdiff x)

runPercentageCheck :: String -> Int -> Maybe Double -> Maybe [String] -> Arguments -> Wordpairs Double 
runPercentageCheck inputdata wordlen distorall blist arg = case distorall of
    Nothing -> instring
    Just i -> filterWordpairsByDistance instring i
  where instring = sortWordsByString . filterMatchingWords $ dwords
        dwords = checkWordList fwords wordlen
        fwords = runFilterFlags cwords arg blist
        cwords = zipWords inputdata "percentage"

processHumanPercentageData :: Wordpairs Double -> [String]
processHumanPercentageData [] = ["No (more) matches found"]
processHumanPercentageData (x:xs) = ("\'" ++ word ++ "\'"
                         ++ " at coordinates "
                         ++ coordinates
                         ++ " with an intervening distance of %"
                         ++ distance') : processHumanPercentageData xs
                         where word = getWordPairString x
                               coordinates = show (getWordpairCoords x)
                               distance' = take 7 (show (pdiff x))

processMachinePercentageData :: Wordpairs Double -> IO ()
processMachinePercentageData x = printBox $ hsep 2 left (map (vcat left . map text) 
                                    (transpose $ ["Coord1", "Coord2", "Word", "Distance"]:processMachinePercentageData' x))

processMachinePercentageData' :: Wordpairs Double -> [[String]]
processMachinePercentageData' [] = []
processMachinePercentageData' (x:xs) = words (coordinates1 coordinates'
                                    ++ " "
                                    ++ coordinates2 coordinates'
                                    ++ " "
                                    ++ word
                                    ++ " "
                                    ++ distance') : processMachinePercentageData' xs
                                      where word = getWordPairString x
                                            coordinates' = getWordpairCoords x
                                            coordinates1 ((r1,s1),(_,_)) = show r1 ++ "," ++ show s1
                                            coordinates2 ((_,_),(r2,s2)) = show r2 ++ "," ++ show s2
                                            distance' = take 7 $ show (pdiff x)

processDataVim :: (NumOps a) => Wordpairs a -> String
processDataVim [] = ""
processDataVim (x:xs) = ("lnum=" ++ "\'" ++ linum1 ++ "\', "  ++ "col=" ++ "\'" ++ colnum1 ++ "\', " ++ "text=" ++ "\'" ++ word ++ "\'\n") ++
                        ("lnum=" ++ "\'" ++ linum2 ++ "\', "  ++ "col=" ++ "\'" ++ colnum2 ++ "\', " ++ "text=" ++ "\'" ++ word ++ "\'\n") ++
                        processDataVim xs
                         where word = getWordPairString x
                               coordinates' = getWordpairCoords x
                               coordinates1 ((r1,_),(_,_)) = show r1
                               coordinates1' ((_,_),(r1,_)) = show r1
                               linum1 = coordinates1 coordinates'
                               linum2 = coordinates1' coordinates'
                               coordinates2 ((_,s1),(_,_)) = show s1 
                               coordinates2' ((_,_),(_,s1)) = show s1 
                               colnum1 = coordinates2 coordinates'
                               colnum2 = coordinates2' coordinates'
                               
-- Function that provides summary totals when -h flag is passed
summaryData :: Int -> Int -> Int -> IO()
summaryData x y z = do
            putStrLn ""
            putStrLn $ "Found " ++ show x  ++ " pairs of words "
              ++ show y ++ " or more characters in length out of "
              ++ show z ++ " words total."
